src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/NextReviewDateUpdaterService.kt

```diff
@@ -84,6 +87,25 @@ class NextReviewDateUpdaterService(
       )
     }
 
-    return nextReviewDateRepository.saveAll(nextReviewDateRecords).toList().toMapByBookingId()
+    val nextReviewDatesAfterUpdate: Map<Long, LocalDate> = nextReviewDateRepository.saveAll(nextReviewDateRecords).toList().toMapByBookingId()
+
+    // Determine which next review dates records actually changed
+    val bookingIdsChanged = bookingIds.filter { bookingId ->
+      nextReviewDatesBeforeUpdate[bookingId] != null && // only publish domain events when next review date changed
+        nextReviewDatesAfterUpdate[bookingId] != nextReviewDatesBeforeUpdate[bookingId]
+    }
+
+    // TODO: It should be possible to publish all these domain events concurrently rather than sequentially
+    bookingIdsChanged.forEach { bookingId ->
+      snsService.publishDomainEvent(
+        id = bookingId,
+        nomsNumber = offendersMap[bookingId]!!.prisonerNumber,
+        occurredAt = LocalDateTime.now(clock),
+        eventType = IncentivesDomainEventType.PRISONER_NEXT_REVIEW_DATE_CHANGED,
+        description = "A prisoner next review date has changed",
```

Comment:
maybe "A prisoner's next _incentive_ review date has changed"

Reply:
I'll change as it's good to give more context 👍 (it's kinda implicit because of the full domain event type but it's good to be more explicit nevertheless)

Reply:
the whole description is implicit in the event name, but i figure the description might as well be a sentence to explain it

Reply:
Done here: https://github.com/ministryofjustice/hmpps-incentives-api/pull/295/commits/7ce33374de254b32ad57008e0fd114a060b4949d